Scripts- PowerShell and MS.Graph Module/W365 - List CloudPCs by OPNC-Hosted.ps1

```diff
@@ -0,0 +1,95 @@
+#>
+
+#Function to check if MS.Graph module is installed and is current version
+function invoke-graphmodule {
+    $graphavailable = (find-module -name microsoft.graph)
+    $vertemp = $graphavailable.version.ToString()
+    Write-Output "Latest version of Microsoft.Graph module is $vertemp" | out-host
+    $graphcurrent = (get-installedmodule -name Microsoft.Graph -ErrorAction SilentlyContinue) 
+
+    if ($graphcurrent -eq $null) {
+        write-output "Microsoft.Graph module is not installed. Installing..." | out-host
+        try {
+            Install-Module Microsoft.Graph -Force -ErrorAction Stop
+        }
+        catch {
+            write-output "Failed to install Microsoft.Graph Module" | out-host
+            write-output $_.Exception.Message | out-host
+            Return 1
+        }
+    }
+    $graphcurrent = (get-installedmodule -name Microsoft.Graph)
+    $vertemp = $graphcurrent.Version.ToString() 
+    write-output "Current installed version of Microsoft.Graph module is $vertemp" | out-host
+
+    if ($graphavailable.Version -gt $graphcurrent.Version) { write-host "There is an update to this module available." }
+    else
+    { write-output "The installed Microsoft.Graph module is up to date." | out-host }
+}
+
+#Function to connect to the MS.Graph PowerShell Enterprise App
+function connect-msgraph {
+
+    $tenant = get-mgcontext
+    if ($tenant.TenantId -eq $null) {
+        write-output "Not connected to MS Graph. Connecting..." | out-host
+        try {
+            Connect-MgGraph -Scopes "CloudPC.Read.All" -ErrorAction Stop | Out-Null
+        }
+        catch {
+            write-output "Failed to connect to MS Graph" | out-host
+            write-output $_.Exception.Message | out-host
+            Return 1
+        }   
+    }
+    $tenant = get-mgcontext
+    $text = "Tenant ID is " + $tenant.TenantId
+    Write-Output "Connected to Microsoft Graph" | out-host
+    Write-Output $text | out-host
+}
+#Function to set the profile to beta
+function set-profile {
+    Write-Output "Setting profile as beta..." | Out-Host
+    Select-MgProfile -Name beta
+}
+
+#Commands to load MS.Graph modules
+if (invoke-graphmodule -eq 1) {
+    write-output "Invoking Graph failed. Exiting..." | out-host
+    Return 1
+}
+
+#Command to connect to MS.Graph PowerShell app
+if (connect-msgraph -eq 1) {
+    write-output "Connecting to Graph failed. Exiting..." | out-host
+    Return 1
+}
+
+set-profile
+
+$OPNCs = Get-MgDeviceManagementVirtualEndpointOnPremisesConnection
+$CloudPCs = Get-MgDeviceManagementVirtualEndpointCloudPC
+
+foreach ($OPNC in $OPNCs) {
+    Write-Host ""
+    Write-Host $OPNC.DisplayName
+    foreach ($CloudPC in $CloudPCs) {
+        if ($CloudPC.OnPremisesConnectionName -eq $OPNC.DisplayName) {
+            Write-Host $CloudPC.DisplayName
+        }
+    }
+}
+
+Write-Host ""
+Write-Host "Microsoft Hosted Network"
+foreach ($CloudPC in $CloudPCs) {
+    if ($CloudPC.OnPremisesConnectionName -eq "") {
```

Comment:
Why equal a null string? What's the usage of this code snippet? Do you plan to use this to list the CloudPCs which OPNC name is null?

Reply:
I am using the null value to differentiate between systems that have an OPNC and ones on the MS hosted network. As systems on the MS hosted network do not have a value in the opnc, it seemed like a safe way to tell the difference